arrows/vxl/average_frames.cxx

```diff
@@ -0,0 +1,788 @@
+{
+public:
+  exponential_frame_averager( bool const should_round = false,
+                              double const new_frame_weight = 0.5 );
+
+  virtual ~exponential_frame_averager() {}
+
+  /// Process a new frame, returning the current frame average.
+  virtual void process_frame( vil_image_view< PixType > const& input,
+                              vil_image_view< PixType >& average );
+
+  /// Reset the internal average.
+  virtual void reset();
+
+private:
+  /// The exponential averaging coefficient
+  double new_frame_weight_;
+
+  /// The number of observed frames since the last reset
+  unsigned frame_count_;
+};
+
+// ---------------------------------------------------------------------------
+/// A windowed frame averager
+template < typename PixType >
+class windowed_frame_averager : public online_frame_averager< PixType >
+{
+public:
+  using input_type = vil_image_view< PixType >;
+
+  windowed_frame_averager( bool const should_round = false,
+                           unsigned const window_length = 20 );
+
+  virtual ~windowed_frame_averager() {}
+
+  /// Process a new frame, returning the current frame average.
+  virtual void process_frame( vil_image_view< PixType > const& input,
+                              vil_image_view< PixType >& average );
+
+  /// Reset the internal average.
+  virtual void reset();
+
+  /// Get number of frames used in the current window
+  virtual size_t frame_count() const;
+
+private:
+  /// Buffer containing pointers to last window_length frames
+  std::deque< vil_image_view< PixType > > window_buffer_;
+  size_t window_buffer_capacity_;
+};
+
+// ----------------------------------------------------------------------------
+// Shared functionality - process a frame while computing variance
+template < typename PixType >
+void
+online_frame_averager< PixType >
+::process_frame( vil_image_view< PixType > const& input,
+                 vil_image_view< PixType >& average,
+                 vil_image_view< double >& variance )
+{
+  // Check if this is the first time we have processed a frame of this size
+  if( dev1_tmp_space_.ni() != input.ni() ||
+      dev1_tmp_space_.nj() != input.nj() ||
+      dev1_tmp_space_.nplanes() != input.nplanes() )
+  {
+    dev1_tmp_space_.set_size( input.ni(), input.nj(), input.nplanes() );
+    dev2_tmp_space_.set_size( input.ni(), input.nj(), input.nplanes() );
+    variance.set_size( input.ni(), input.nj(), input.nplanes() );
+    variance.fill( 0.0 );
+    this->process_frame( input, average );
+    return;
+  }
+
+  // Calculate difference from last average
+  vil_math_image_abs_difference( input, this->last_average_, dev1_tmp_space_ );
+
+  // Update internal average
+  this->process_frame( input, average );
+
+  // Update the variance
+  vil_math_image_abs_difference( input, average, dev2_tmp_space_ );
+  vil_math_image_product( dev1_tmp_space_, dev2_tmp_space_, dev1_tmp_space_ );
+  variance.deep_copy( dev1_tmp_space_ );
```

Comment:
I think `dev1_tmp_space_` holds a running estimate of the variance. So we need to update it, in addition to returning the value by reference.

On another note, I think we've agreed the variance is improperly computed, at least in the general case. However, since it appears to be a useful feature, we're going to leave it.

Reply:
I don't see it being used anywhere except here, as scratch space. Only the line above ever reads from it, and we overwrote it a couple lines above that. The name does rather suggest it is only scratch space...

Reply:
Hmm yeah I'm not sure what's going on. I'm guessing variance calculation is even more broken than before then. I'd say we merge it and then I'll open a follow up PR to deal with variance.